tree_sitter/binding.c

```diff
@@ -878,7 +1085,130 @@ static PyObject *query_new_internal(
+    PyList_SetItem(query->string_values, i, PyUnicode_FromStringAndSize(string_value, length));
+  }
+
+  unsigned pattern_count = ts_query_pattern_count(query->query);
+  query->text_predicates = PyList_New(pattern_count);
+  if (query->text_predicates == NULL) {
+    PyErr_SetString(PyExc_RuntimeError, "Failed to allocate list for query->text_predicates");
+    goto error;
+  }
+
+  for (unsigned i = 0; i < pattern_count; i++) {
+    unsigned length;
+    const TSQueryPredicateStep* predicate_step = ts_query_predicates_for_pattern(query->query, i, &length);
+    pattern_text_predicates = PyList_New(0);
+    if (pattern_text_predicates == NULL) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to allocate list for pattern_text_predicates");
+      goto error;
+    }
+    for (unsigned j = 0; j < length; j++) {
+      unsigned predicate_len = 0;
+      while ((predicate_step + predicate_len)->type != TSQueryPredicateStepTypeDone)
+        predicate_len++;
+
+      if (predicate_step->type != TSQueryPredicateStepTypeString) {
+        PyErr_SetString(PyExc_RuntimeError, "Capture predicate must start with a string");
+        goto error;
+      }
+
+      // Build a predicate for each of the supported predicate function names
+      PyObject *operator_name = PyList_GetItem(query->string_values, predicate_step->value_id);
+      if (PyUnicode_CompareWithASCIIString(operator_name, "eq?") == 0 ||
+          PyUnicode_CompareWithASCIIString(operator_name, "not-eq?") == 0) {
+        if (predicate_len != 3) {
+          PyErr_SetString(PyExc_RuntimeError, "Wrong number of arguments to #eq? or #not-eq? predicate");
+          goto error;
+        }
+        if ((predicate_step + 1)->type != TSQueryPredicateStepTypeCapture) {
+          PyErr_SetString(PyExc_RuntimeError, "First argument to #eq? or #not-eq? must be a capture name");
+          goto error;
+        }
+        int is_positive = PyUnicode_CompareWithASCIIString(operator_name, "eq?") == 0;
+        switch ((predicate_step + 2)->type) {
+          case TSQueryPredicateStepTypeCapture:
+            ;
+            CaptureEqCapture *capture_eq_capture_predicate = (CaptureEqCapture *)capture_eq_capture_new_internal(
+              (predicate_step+1)->value_id, (predicate_step+2)->value_id, is_positive
+            );
+            if (capture_eq_capture_predicate == NULL) {
+              PyErr_SetString(PyExc_RuntimeError, "Failed to allocate CaptureEqCapture text predicate");
+              goto error;
+            }
+            PyList_Append(pattern_text_predicates, (PyObject *)capture_eq_capture_predicate);
+            Py_DECREF(capture_eq_capture_predicate);
+            break;
+          case TSQueryPredicateStepTypeString:
+            ;
+            Py_ssize_t size;
+            const char *string_value = PyUnicode_AsUTF8AndSize(
+              PyList_GetItem(query->string_values, (predicate_step+2)->value_id), &size
+            );
```

Comment:
The main reason of stashing the strings in the `string_values` list was that I saw that something similar was done for capture names (stored in `query->capture_names`), and that the Rust bindings also created the `string_values` list instead of calling `ts_query_string_value_for_id` repeatedly. So I guessed this approach would create less friction with existing code. But yeah, it's kind of superfluous, so no problem removing it if @maxbrunsfeld is fine with that.

Reply:
That might be due to rust wanting more clearly defined lifetimes for the strings - here it ends up kind of awkward because we don't even want the string as a `PyObject` at any point.

For `capture_names`, that list is filled in `query_new_internal` but fetched in `query_captures` and is ultimately returned to Python-land as a string, so holding onto a list has utility there.

With `string_values`, the actual objects are only used in the one function, and never returned to Python, so the utility of holding it in a Python object is very limited (and it's still confusing that we're passing around long-lived pointers to the utf8 cache buffer of the string PyObject)

Reply:
If we do the regex matching in Python though, wouldn't we end up reporting the string values as Python strings? Or are you thinking that we'd actually use the Python `re` module  entirely from C? To me, it seemed simpler to deal with the regexes in the Python code, but I'm open to either approach.

Reply:
Ok, I think we should store a `PyObject *` on the `CaptureMatchString` and `CaptureEqString`.

For `MatchString`, I think it should be a compiled regular expression from `re.compile`.

For `CaptureEqString`, I think it needs to be a PyObject of `bytes` rather than a `str`, so we can compare directly to the `bytes` we get out of `node_get_text` (which is currently used with `strcmp` via `PyBytes_AsString`).

`string_values` should only be created if tree-sitter will often in practice reuse a lot of IDs, such that the list allocation is less impactful than the bytes allocations. In that case, it still doesn't need to be stored beyond the initializer, and can be a local that is released at the end.

Reply:
Fixed.